Source/Plugin.BLE.iOS/Adapter.cs

```diff
@@ -366,5 +365,141 @@ public static List<AdvertisementRecord> ParseAdvertismentData(NSDictionary adver
+                        return device;
+                    }
+                }
+            }
+        }
+
+        private async Task<IDevice> ScanAndConnect(string friendlyName, Func<IDevice, bool> deviceFilter)
+        {
+            var peripheral = await ScanForPeripheralAsync(friendlyName, deviceFilter);
+            return await TryToConnectAsync(peripheral);
+        }
+
+        private async Task<CBPeripheral> ScanForPeripheralAsync(string friendlyName, Func<IDevice, bool> deviceFilter)
+        {
+            var taskCompletionSource = new TaskCompletionSource<CBPeripheral>();
+            var stopToken = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+            EventHandler<DeviceEventArgs> handler = (sender, args) =>
+            {
+                var peripheral = args.Device.NativeDevice as CBPeripheral;
+
+                if(peripheral.Name == friendlyName)
+                {
+                    stopToken.Cancel();
+                    taskCompletionSource.SetResult(peripheral);
+                }
+            };
+
+            try
+            {
+                DeviceDiscovered += handler;
+                await StartScanningForDevicesAsync(
+                    deviceFilter: deviceFilter,
+                    cancellationToken: stopToken.Token);
+                return await taskCompletionSource.Task;
+            }
+            catch(OperationCanceledException)
+            {
+                return null;
+            }
+            finally
+            {
+                DeviceDiscovered -= handler;
+            }
+        }
+
+        /// <summary>
+        /// A known peripheral is either one we can find by uuid or one we're already connected to
+        /// </summary>
+        private CBPeripheral TryToRetrieveKnownPeripheral(NSUuid nsuuid)
+        {
+            var peripherals = _centralManager.RetrievePeripheralsWithIdentifiers(nsuuid);
+            var peripheral = peripherals.SingleOrDefault();
+            if (peripheral == null)
+            {
+                var connectedPeripherals = _centralManager.RetrieveConnectedPeripherals(new CBUUID[0]);
+                var cbuuid = CBUUID.FromNSUuid(nsuuid);
+                peripheral = connectedPeripherals.SingleOrDefault(p => p.UUID.Equals(cbuuid));
+            }
+
+            return peripheral;
+        }
+
+        private async Task<IDevice> TryToConnectAsync(CBPeripheral peripheral)
+        {
+            if(peripheral == null)
+            {
+                return null;
+            }
+
+            var completionSource = new TaskCompletionSource<IDevice>();
+            EventHandler<CBPeripheralEventArgs> connectedEvent = (sender, args) =>
+            {
+                var device = new Device(this, args.Peripheral);
+                completionSource.TrySetResult(device);
+            };
+
+            EventHandler<CBPeripheralErrorEventArgs> errorEvent = (sender, args) =>
+            {
+                completionSource.TrySetResult(null);
```

Comment:
Maybe the NSError should be pulled out from args and logged? And what would happen if we use TrySetException instead?

Reply:
Agreed we should log it. 
I do not think exceptions are appropriate, since it is an expected state that we do not find anything. We should avoid doing flow control using exceptions, which would happen in that case. Or am I missing something obvious?